crates/rspack_binding_api/src/normal_module_factory.rs

```diff
@@ -51,20 +50,107 @@ pub struct JsCreateData {
   pub resource: String,
 }
 
-#[derive(Serialize)]
+impl From<&NormalModuleCreateData> for JsCreateData {
+  fn from(value: &NormalModuleCreateData) -> Self {
+    Self {
+      request: value.request.to_owned(),
+      user_request: value.user_request.to_owned(),
+      resource: value.resource_resolve_data.resource.to_owned(),
+    }
+  }
+}
+
+impl JsCreateData {
+  pub fn update_nmf_data(self, create_data: &mut NormalModuleCreateData) {
+    fn update_resource_data(old_resource_data: &mut ResourceData, new_resource: String) {
+      if old_resource_data.resource_path.is_some()
+        && let Some(parsed) = parse_resource(&new_resource)
+      {
+        old_resource_data.set_path(parsed.path);
+        old_resource_data.set_query_optional(parsed.query);
+        old_resource_data.set_fragment_optional(parsed.fragment);
+      }
+      old_resource_data.set_resource(new_resource);
+    }
+
+    create_data.request = self.request;
+    create_data.user_request = self.user_request;
+    if create_data.resource_resolve_data.resource != self.resource {
+      update_resource_data(&mut create_data.resource_resolve_data, self.resource);
+    }
+  }
+}
+
+#[derive(Debug, Serialize)]
 #[serde(rename_all = "camelCase")]
-pub struct JsAfterResolveData {
+#[napi(object)]
+pub struct JsResolveData {
   pub request: String,
   pub context: String,
-  pub issuer: String,
-  pub issuer_layer: Option<String>,
+  pub context_info: ContextInfo,
   pub file_dependencies: Vec<String>,
   pub context_dependencies: Vec<String>,
   pub missing_dependencies: Vec<String>,
   pub create_data: Option<JsCreateData>,
 }
 
-pub type JsAfterResolveOutput = (Option<bool>, Option<JsCreateData>);
+impl JsResolveData {
+  pub fn from_nmf_data(
+    data: &ModuleFactoryCreateData,
+    create_data: Option<&NormalModuleCreateData>,
+  ) -> Self {
+    JsResolveData {
+      request: data.request.to_string(),
+      context: data.context.to_string(),
+      context_info: ContextInfo {
+        issuer: data
+          .issuer
+          .as_ref()
+          .map(|issuer| issuer.to_string())
+          .unwrap_or_default(),
+        issuer_layer: data.issuer_layer.clone(),
+      },
+      file_dependencies: data
+        .file_dependencies
+        .clone()
+        .into_iter()
+        .map(|item| item.to_string_lossy().to_string())
```

Comment:
```suggestion
        .map(|item| item.to_string_lossy().into_owned())
```

Reply:
It would be better to handle this error if possible, because if lossy occurs, it is unlikely to be expected behavior.